chaos/cluster.go

```diff
@@ -0,0 +1,264 @@
+	errDone := make(chan struct{})
+	go func() {
+		for err := range errCh {
+			errs = append(errs, err)
+		}
+		errDone <- struct{}{}
+	}()
+
+	parallel := make(chan struct{}, maxParallel)
+	glog.V(4).Infof("parallel reboots: %d", maxParallel)
+	var wg sync.WaitGroup
+	for i := range hosts {
+		wg.Add(1)
+		parallel <- struct{}{}
+		go func(host string) {
+			defer wg.Done()
+			if err := cl.RebootNode(host); err != nil {
+				errCh <- err
+			}
+			<-parallel
+		}(hosts[i])
+	}
+	wg.Wait()
+	close(errCh)
+	close(parallel)
+	<-errDone
+
+	return errors.NewAggregate(errs)
+}
+
+// RebootNode reboots a node addressable with `host`.
+// Uses *Cluster sshClient and RebootAfter.
+func (cl *Cluster) RebootNode(host string) error {
+	glog.V(4).Infof("node: %s starting reboot", host)
+	// disable kubelet if RebootAfter > 0
+	if cl.RebootAfter.Seconds() > 0 {
+		glog.V(4).Infof("node: %s trying to disable kubelet", host)
+		stdout, stderr, err := cl.sshClient.SSH(host, kubeletDisableCmd)
+		if err != nil {
+			return fmt.Errorf("node: %s issuing kubelet disable command failed\nstdout:%s\nstderr:%s", host, stdout, stderr)
+		}
+		glog.V(4).Infof("node: %s disabled kubelet", host)
+	}
+
+	// reboot node
+	stdout, stderr, err := cl.sshClient.SSH(host, rebootCmd)
+	if _, ok := err.(*ssh.ExitMissingError); ok {
+		// A terminated session is perfectly normal during reboot.
+		err = nil
+	}
+	if err != nil {
+		return fmt.Errorf("node: %s issuing reboot command failed\nstdout:%s\nstderr:%s", host, stdout, stderr)
+	}
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+		stdout, stderr, err = cl.sshClient.SSH(host, systemUpCmd)
+		if err != nil {
+			glog.Errorf("node: %s %v: %v", host, err, stderr)
+			return false, nil
+		}
+		if !bytes.Contains(stdout, []byte("active")) {
+			glog.Errorf("node: %s system is not running yet", host)
+			return false, nil
+		}
+		return true, nil
+
+	}); err != nil {
+		return fmt.Errorf("node: %s reboot failed: %v", host, err)
+	}
+
+	// wait for RebootAfter before starting kubelet again.
+	if cl.RebootAfter.Seconds() > 0 {
+		w := time.After(cl.RebootAfter)
+		glog.V(4).Infof("node: %s waiting for %s before starting kubelet", host, cl.RebootAfter)
+		<-w
+		stdout, stderr, err := cl.sshClient.SSH(host, kubeletStartCmd)
+		if err != nil {
+			return fmt.Errorf("node: %s issuing kubelet start command failed\nstdout:%s\nstderr:%s", host, stdout, stderr)
+		}
+		glog.V(4).Infof("node: %s kubelet started", host)
+	}
+
+	glog.V(4).Infof("node: %s reboot successful", host)
+	return nil
+}
+
+// IsMaster returns true if the node's labels contains "node-role.kubernetes.io/master".
+func IsMaster(n *v1.Node) bool {
+	_, ok := n.Labels[NodeRoleMasterLabel]
+	return ok
+}
+
+// IsWorker returns true if the node's labels contains "node-role.kubernetes.io/node".
+func IsWorker(n *v1.Node) bool {
+	_, ok := n.Labels[NodeRoleWorkerLabel]
+	return ok
+}
+
+func hostsFromNodes(in []*v1.Node) (out []string) {
```

Comment:
nodes []*v1.Node hosts []string

in, out has zero meaning. we should avoid naming like that. (i know k8s uses this naming a lot, but it is a bad practice.)